toponetx/classes/path.py

```diff
@@ -0,0 +1,227 @@
+            List of elementary p-path objects representing the boundaries.
+
+        Notes
+        -----
+        - In order to construct a set of boundary (p-1)-paths of an elementary p-path (n1, ..., np), we iteratively omit one node at a time from
+        the elementary p-path. The resulted boundary paths are elementary (p-1)-paths in the form of (n1, ..., ni-1, ni+1, ..., np), where i is
+        the index of the omitted node. However, some of the elementary (p-1)-paths may be non-existing (or non-allowed in our context), so we
+        discard them from the boundary set.
+
+        Examples
+        --------
+        >>> Path.construct_path_boundaries((1, 2, 3), reserve_sequence_order=False)
+        [(2, 3), (1, 3), (1, 2)]
+        >>> Path.construct_path_boundaries((1, 2, 3), reserve_sequence_order=False, allowed_paths=[(1, 2), (2, 3)])
+        [(2, 3), (1, 2)]
+        """
+        boundaries = list()
+        for i in range(len(elements)):
+            boundary = list(elements[0:i] + elements[(i + 1) :])
+            if (
+                not reserve_sequence_order
+                and len(boundary) > 1
+                and str(boundary[0]) > str(boundary[-1])
+            ):
+                boundary.reverse()
+            if allowed_paths is None or tuple(boundary) in allowed_paths:
+                boundaries.append(tuple(boundary))
+        return boundaries
+
+    @property
+    def boundary(self) -> list[tuple[Hashable]]:
+        """
+        Return list of elementary (p-1)-path objects representing the elementary (p-1)-path on the boundary of the target elementary p-path.
+
+        Returns
+        -------
+        boundaries : list[tuple[Hashable]]
+            List of elementary p-path objects representing the boundary (p-1)-paths.
+        """
+        return self._boundaries
+
+    def clone(self) -> "Path":
+        """
+        Return a shallow copy of the elementary p-path.
+
+        Returns
+        -------
+        Path
+            A shallow copy of the elementary p-path.
+        """
+        return Path(
+            self.elements,
+            name=self.name,
+            construct_boundaries=self.construct_boundaries,
+            **self._properties,
+        )
+
+    def __check_inputs(self, elements: Any, reserve_sequence_order: bool):
+        """
+        Sanity check for inputs, as sequence order matters.
+
+        Parameters
+        ----------
+        elements: Any
+            The nodes in the elementary p-path.
+        reserve_sequence_order : bool
+            If True, reserve the order of the sub-sequence of nodes in the elementary p-path.
+            Else, the sub-sequence of nodes in the elementary p-path will be reversed if the first index is larger than the last index.
+            This variable is only used for sanity check on the input.
+        """
+        if isinstance(elements, Sequence):
+            if not isinstance(elements, list) and not isinstance(elements, tuple):
```

Comment:
Does it have to be a `list` or `tuple`? I would prefer to use the broadest possible protocol in `collections.abc`.

Reply:
I think paths have to be ordered, so sets or frozensets do not work.

Reply:
There are many other data structures, that's why the protocols in `collections.abc` exist. I don't see why e.g., `collections.abc.Sequence` wouldn't be enough? All sequences are ordered.

Reply:
Yes you're correct on this. `collections.abc.Sequence` is sufficient.